File: julia/single_dose.jl
"""
Single Cigarette Response - Julia Implementation

Demonstrates acute pharmacological response to a single bolus dose.
"""

include("NicotineModel.jl")
using Plots

println("="^60)
println("EXAMPLE: SINGLE CIGARETTE DOSE (Julia)")
println("="^60)

# Create model parameters
params = Parameters()

# Single cigarette bolus
I = bolus_input(5.0, 10.0, 5.0)

# Simulate
println("\nSimulating 5-hour period...")
tspan = (0.0, 300.0)
sol = simulate(tspan, I, params)

# Extract results
t = sol.t
N = [u[1] for u in sol.u]
Ra = [u[2] for u in sol.u]
Rd = [u[3] for u in sol.u]
RT = [u[4] for u in sol.u]
D = [u[5] for u in sol.u]

# Analysis
println("\nPeak responses:")
println("  Nicotine:  $(maximum(N)) μM")
println("  Active receptors: $(maximum(Ra))")
println("  Dopamine: $(maximum(D)) AU")

# Time to peak
t_peak_D = t[argmax(D)]
println("\nTime to peak dopamine: $(round(t_peak_D, digits=1)) min")

# Nicotine half-life
N_peak = maximum(N)
idx_peak = argmax(N)
idx_half = idx_peak + findfirst(N[idx_peak:end] .< N_peak/2)
if !isnothing(idx_half)
    t_half = t[idx_half] - t[idx_peak]
    println("Nicotine half-life: $(round(t_half, digits=1)) min")
end

# Visualization
println("\nGenerating plots...")

p = plot(layout=(4,1), size=(1000, 800), legend=:topright)

# Nicotine
plot!(p[1], t, N, linewidth=2, color=:blue, label="Nicotine", 
     ylabel="Nicotine [μM]", grid=true, title="FAST: Nicotine Kinetics")

# Active & desensitised receptors
plot!(p[2], t, Ra, linewidth=2, color=:green, label="Active", 
     ylabel="Receptor Fraction", grid=true)
plot!(p[2], t, Rd, linewidth=2, color=:red, linestyle=:dash, label="Desensitised",
     title="MEDIUM: Receptor Dynamics")

# Total receptors
plot!(p[3], t, RT, linewidth=2, color=:orange, label="R_T",
     ylabel="Total Receptors", grid=true, title="SLOW: Receptor Pool")

# Dopamine
plot!(p[4], t, D, linewidth=2, color=:purple, label="Dopamine",
     ylabel="Dopamine [AU]", xlabel="Time [min]", grid=true,
     title="FAST: Dopamine Response")

plot!(p, plot_title="Single Cigarette: Acute Response", 
     plot_titlefontsize=14)

savefig(p, "single_dose_julia.png")
println("✓ Saved: single_dose_julia.png")

# Stability analysis
println("\n" * "="^60)
println("STABILITY ANALYSIS")
println("="^60)

u_final = sol.u[end]
eigenvals, timescales, is_stable = stability_analysis(u_final, 0.0, params)

println("\nFinal state stability: $(is_stable ? "STABLE" : "UNSTABLE")")
println("Eigenvalues (real parts): $(round.(real.(eigenvals), digits=6))")

valid_ts = filter(!isnan, timescales)
if !isempty(valid_ts)
    println("\nTimescales:")
    for (i, τ) in enumerate(sort(valid_ts))
        println("  τ_$i = $(round(τ, digits=2)) min")
    end
end

ts_results = timescale_separation(u_final, 0.0, params)
println("\nTimescale separation: $(round(ts_results.ratio, digits=0))×")

println("\n" * "="^60)
println("KEY OBSERVATIONS")
println("="^60)
println("1. Fast nicotine clearance (~120 min half-life)")
println("2. Transient dopamine surge (reinforcement window)")
println("3. Receptor desensitisation develops and recovers")
println("4. Complete return to baseline (no long-term adaptation)")
println("5. System is stable throughout")
println("="^60)

File: examples/single_dose.py
"""
Example: Single cigarette dose response.

Demonstrates:
- Fast nicotine kinetics
- Transient receptor activation
- Dopamine surge
- Return to baseline
"""

import sys
sys.path.append('../src')

from model import NicotineRewardModel, bolus_input
from visualization import Visualizer
import matplotlib.pyplot as plt


def main():
    print("=" * 60)
    print("EXAMPLE: SINGLE CIGARETTE DOSE")
    print("=" * 60)
    
    # Create model
    model = NicotineRewardModel()
    
    # Single cigarette: rapid bolus delivery
    # Typical cigarette delivers ~1-2 mg nicotine, peak brain conc ~0.2-0.5 μM
    I = bolus_input(
        amplitude=5.0,  # Fast input rate during smoking
        t_start=10,     # Starts at 10 min
        duration=5      # Smoking duration ~5 min
    )
    
    # Simulate for 5 hours
    print("\nSimulating 5-hour period...")
    result = model.simulate(
        t_span=(0, 300),
        input_func=I
    )
    
    # Analysis
    print("\nPeak responses:")
    print(f"  Nicotine:  {result['N'].max():.3f} μM")
    print(f"  Active receptors: {result['R_a'].max():.3f}")
    print(f"  Dopamine: {result['D'].max():.3f} AU")
    
    # Time to peak
    t_peak_D = result['t'][result['D'].argmax()]
    print(f"\nTime to peak dopamine: {t_peak_D:.1f} min")
    
    # Half-life (nicotine clearance)
    N_peak = result['N'].max()
    idx_peak = result['N'].argmax()
    idx_half = idx_peak + (result['N'][idx_peak:] < N_peak/2).argmax()
    t_half = result['t'][idx_half] - result['t'][idx_peak]
    print(f"Nicotine half-life: {t_half:.1f} min")
    
    # Visualization
    print("\nGenerating plots...")
    vis = Visualizer()
    
    # Time series
    fig1 = vis.plot_timeseries(
        result,
        variables=['N', 'R_a', 'R_d', 'D'],
        title='Single Cigarette: Acute Response',
        save_path='single_dose_timeseries.png'
    )
    
    # Phase portrait
    fig2 = vis.plot_phase_portrait_2d(
        result,
        var1='R_a',
        var2='D',
        title='Phase Portrait: Receptor Activation → Dopamine',
        save_path='single_dose_phase.png'
    )
    
    print("\n✓ Plots saved:")
    print("  - single_dose_timeseries.png")
    print("  - single_dose_phase.png")
    
    plt.show()
    
    print("\n" + "=" * 60)
    print("Key observations:")
    print("=" * 60)
    print("1. Fast nicotine kinetics (~2h half-life)")
    print("2. Rapid dopamine surge (reinforcement window)")
    print("3. Receptor desensitisation develops transiently")
    print("4. Full recovery to baseline (no long-term changes)")
    print("=" * 60)


if __name__ == "__main__":
    main()
